src/backends/policy_engine/souffle/check_policy_compliance.cc

```diff
@@ -69,29 +72,60 @@ int main(int argc, char* argv[]) {
   absl::SetProgramUsageMessage(kUsageMessage);
   absl::ParseCommandLine(argc, argv);
 
+  // Read the sql policy rules file.
+  absl::StatusOr<std::string> sql_policy_rules =
+          ReadFileContents(absl::GetFlag(FLAGS_sql_policy_rules));
+  if (!sql_policy_rules.ok()) {
+      LOG(ERROR) << "Error reading sql policy rules file: "
+                 << sql_policy_rules.status();
+      return 2;
+  }
+
+  // TODO(cromwellian): these flags are mutually exclusive, add a check
   // Parse the given IR file.
-  absl::StatusOr<std::string> ir_string =
-      ReadFileContents(absl::GetFlag(FLAGS_ir));
-  if (!ir_string.ok()) {
-    LOG(ERROR) << "Error reading IR file: " << ir_string.status();
-    return 2;
+  auto ir_path = absl::GetFlag(FLAGS_ir);
+  auto proto_path = absl::GetFlag(FLAGS_proto);
+
+  if (ir_path.has_value() && proto_path.has_value()) {
+      LOG(ERROR) << "Both --ir and --proto cannot be used simultaneously.";
+      return 2;
   }
 
-  // Read the sql policy rules file.
-  absl::StatusOr<std::string> sql_policy_rules =
-      ReadFileContents(absl::GetFlag(FLAGS_sql_policy_rules));
-  if (!ir_string.ok()) {
-    LOG(ERROR) << "Error reading sql policy rules file: "
-               << sql_policy_rules.status();
-    return 2;
+  std::unique_ptr<raksha::ir::Module> ir_module;
+  std::unique_ptr<raksha::ir::SsaNames> ir_ssa_names;
+  std::unique_ptr<raksha::ir::IRContext> ir_context;
+
+  if (ir_path.has_value()) {
+      absl::StatusOr<std::string> ir_string = ReadFileContents(*ir_path);
+      if (!ir_string.ok()) {
+          LOG(ERROR) << "Error reading IR file: " << ir_string.status();
+          return 2;
+      }
+
+      IrProgramParser irParser;
+      auto [context, module, ssa_names] = irParser.ParseProgram(*ir_string);
+      ir_module = std::move(module);
+      ir_ssa_names = std::move(ssa_names);
+      ir_context = std::move(context);
```

Comment:
I'm making these moves here because if I don't move ssa_names and context, it'll crash with bad_alloc, even though they aren't used. My guess is when the if{} scope disappears, the local unique_ptr destructs, but there's still some way it's being referenced from module? my C++ fu is too stale.

Reply:
Right, if you don't move them, the ownership is still 'in' the if, so they get cleaned up. Tbh we shouldn't be able to assign to a unique_ptr without a move... but it seems C++ is not as strict as it should be here (*cough* unlike rust *cough*).

Reply:
Yeah, I almost believed the compiler's "auto [x,y,z]" destructuring assignment would do it for us automagically. :(

Reply:
Cool, great suggestions. I had thought of using the ? operator to avoid creating the scoping issue, but for some reason convinced myself the destructuring assignment might not work. :(